### 前情提要

有一个作为消息推送的表，且有一个状态值标识是否推送过。

线上时推送消息宕机了一段时间，导致消息堆积。接口恢复后推送了很多重复消息。

### 实现逻辑

链路：xxl服务端 -》串行调用xxl执行器 -》消息服务 -》消息产商的推送接口

线上问题有两个：

1. xxl服务端宕机
2. 消息产商API超时

推送大致代码

```java
public void pushAll(){
    // 查出要推送的数据
    List<Msg> list = dao.query(...);
    for(Msg msg:list){
        try{
            push(msg);
        } catch(Exption){
            log.error(...);
        }
    }
}

@Transactional(
        propagation = Propagation.REQUIRED,
        isolation = Isolation.DEFAULT,
        timeout = 600,
        rollbackFor = Exception.class)
public void push(Msg msg){
    if(已推送(msg)){
        // 避免重复推送
        return;
    }
    // 标识为已推送
    updateStatus(msg);
    try{
        // 调用产商的推送API
        api.push(msg);
    } catch(Exception e){
        log.error(...);
        throw e;
    }
}
```

### 排查

> 性能不提，祖传代码且不是我负责的模块，不熟，没什么条件自测，改不好容易出事，只是帮忙排查BUG。

首先，`#pushAll()`没有并发限制，并发时会查询出相近的推送列表重复遍历调用`#push()`。

`#push()`问题就多了：

- 虽然有对推送标识做校验，但标识是从`#pushAll()`查询中一次性读取的，并不是库中最新的状态，容易导致重复推送。

- 另一个同事提出产商API超时回滚了推送标识，但实际推送成功，导致后面重复推送。被我排除，`#push()`上的注解`@Transactional`其实并不生效，因为是实例自己的内部调用。本意应该是推送失败时回滚更新标识，但那也应该是`propagation=REQUIRED_NEW`。推送失败依旧会被标识入库，导致丢失。因为`@Transactional`引起了注意，然后排查一遍链路发现，好家伙，前面的方法、接口、类没一个有事务标识，所以这段代码一直运行在无事务的状态。因此排除消息产商超时引起。

转头看看xxl，任务配置咋一看没问题，`单机串行`，似乎并不会导致并发问题。但实际日志中却出现了并发的情况。调用执行器因为消息积压的原因，实际响应时间超过了一分钟，而任务配置的超时时间只有三十秒，导致xxl判定任务超时执行失败。下一个任务在上一个任务实际还没结束，只是响应超时的情况下再次执行了，也就导致了`#pushAll()`的并发。

### 方案

修改`push()`方法：

- 加锁；
- 手动声明新事务；
- 读取最新的推送标识。

还可以用redis key，修改成先推送在更新。

同时负责的同事也提出了一个更快速的方法。xxl任务超时时间调为0，即不超时，同时策略改为`丢弃后续`，这样就能保证xxl同时进行的任务最多仅有一个。有风险，但是是最快、最简单的处理方案。